[DOC]Jupyterhub docs (#4452)
pachyderm/pachyderm: doc/docs/master/how-tos/use-jupyterhub/index.md

One of the most significant advantages of using `python-pachyderm`
is that you can avoid rebuilding your Docker images
every time you update your code.
You can use the `create_python_pipeline` function to create a pipeline from
within your JupyterHub Notebook and store your code locally rather than
in a Docker container.

If you want to create a pipeline from a Docker image or non-Python code, you
can use the
[create_pipeline](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.Client.create_pipeline)
method. This is the method Pachyderm uses when you run `pachctl create
pipeline` or create a pipeline in the Pachyderm UI. When you use
`create_pipeline`, you need to build a new Docker image and push
it to an image registry every
time you update the code in your pipeline. Users that are less familiar
with Docker might find this process a bit cumbersome.

When you use `python-pachyderm`, in addition to the mentioned `create_pipeline` method,
you can use the [create_python_pipeline](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.create_python_pipeline) function that does not require

> you to include your code in a Docker image and rebuild it each time you make

Reviewer: You said this already above. It seem weird to me to talk about fancy create pipeline, then normal create pipeline, then go back to talk more details about fancy. Group things together better? 

e.g Start with: python-pach support the normal create_pipeline method that is available through the CLI, etc etc. 

But it also has an extra abstraction layer that is particularly nice for python users... intorduce fancy create pipeline...

a change. Instead, this function creates a PFS repository
called `<pipeline_name>_source` and puts the source code into it. Also, it

> creates a `<pipeline_name>_build` repository to build Python dependencies.

Reviewer: are there any special things the user needs to do with their requirements.txt file to make this method work or is that req.txt such a python standard that it shouldn't be any issue? @ysimonson

Reviewer 2: It's a pretty common pattern for python projects

# Use JupyterHub with `python-pachyderm`

This section describes how you can use the `python-pachyderm`
client from within the JupyterHub UI.

!!! note

>     You need to have Pachyderm and JupyterHub instlalled on the

Reviewer 2: ```suggestion
    You need to have Pachyderm and JupyterHub installed on the
```

Author: added

    integrated with Pachyderm.

## Overview

JupyterHub is a popular data science platform that enables users
to quickly spin out multiple single-tenant Jupyter Notebook server instances.
Jupyter Notebook provides a Python interactive development environment (IDE)
that is convenient for data science projects. Because of the built-in
rich-text support, visualizations, the easy-to-use web interface, many
enterprise users prefer Jupyter Notebooks to the classic Terminal prompt.
JupyterHub brings all the benefits of Jupyter Notebooks without the need
to install or configure anything on user machines except for a web browser.

[python-pachyderm](https://github.com/pachyderm/python-pachyderm) is an
official Python client for Pachyderm. For Python developers who prefer to
communicate with Pachyderm directly through the API, rather than by using
the `pachctl` tool, `python-pachyderm` is the right choice.
The [API Documentation](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html)
describes various API operations that you can execute to interact with
Pachyderm.

When you deployed JupyterHub by using our deployment script, `python-pachyderm`

> was enabled in JupyterHub so that you can run API requests directly from

Reviewer 2: ```suggestion
was installed in JupyterHub so that you can run API requests directly from
```

Author: added

your Jupyter Notebook.

### Difference in Pipeline Creation Methods

One of the most significant advantages of using `python-pachyderm`

> with JupyterHub is that you can avoid rebuilding your Docker images

Reviewer 2: ```suggestion
is that you can avoid rebuilding your Docker images
```

This convenience has more weight in JupyterHub where the environment is fairly constrained, but it's available to all users of python-pachyderm, regardless of whether they're using JupyterHub.

Author: added

every time you update your code. Because JupyterHub is a Python environment,
you can use the `create_python_pipeline` function to create a pipeline from
within your JupyterHub Notebook and store your code locally rather than

> in a Docker container.

Reviewer 2: This isn't quite true, because JupyterHub can execute non-python code too.

Author: rephrased

> When you use `pachctl` or the Pachyderm UI to create a pipeline, you execute

Reviewer 2: This sentence is confusing. What about something like this?

```markdown
If you want to create a pipeline from a Docker image or non-Python code, you can use
[create_pipeline](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.Client.create_pipeline).
This function is akin to `pachctl create pipeline`.

Author: rephrased a bit.

the
[create_pipeline](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.Client.create_pipeline)
method in the background. This standard workflow requires you to build a new
Docker image and push it to an image registry every time you update the
code in your pipeline. Users that are less familiar with Docker might find this
process a bit cumbersome.

> When you use JupyterHub, in addition to the mentioned `create_pipeline` method,

Reviewer 2: ```suggestion
When you use `python-pachyderm`, in addition to the mentioned `create_pipeline` method,
```

Author: added

you can use the [create_python_pipeline](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.create_python_pipeline) function that does not require
you to include your code in a Docker image and rebuild it each time you make
a change. Instead, this function creates a PFS repository

> called `pipeline_name.source` and puts the source code into it. Also, it

Reviewer 2: ```suggestion
called `<pipeline_name>_source` and puts the source code into it. Also, it
```

Author: added

> creates a `pipeline_name.build` repository to build Python dependencies.

Reviewer 2: ```suggestion
creates a `<pipeline_name>_build` repository to build Python dependencies.
```

Author: added

Therefore, when you use `create_python_pipeline`, your DAG includes two
additional repositories for each pipeline. While you still need a generic
image for your pipeline, your code is stored separately from it.
Because of that, you do not need
to build a new Docker image every time you change something in your

> pipeline code. You can run your code instantly. This method is ideal for

Reviewer 2: While it's true that you need a generic image for the pipeline, I'd leave that bit out since 99% of users will not need to override the default.

Author: removed

# Use JupyterHub with `python-pachyderm`

This section describes how you can use the `python-pachyderm`
client from within the JupyterHub UI.

!!! note
    You need to have Pachyderm and JupyterHub installed on the
    same Kubernetes cluster as described in
    [Deploy Pachyderm with JupyterHub](../../deploy-manage/deploy/deploy-pachyderm-jupyterhub.md).
    Currently, an existing JupyterHub deployment cannot be

>     integrated with Pachyderm.

Reviewer 2: This isn't true, you can modify an existing JupyterHub deployment, and we give some (very incomplete) guides on how to do so in the jupyterhub-pachyderm repo. You just can't use `init.py`, so the steps required are quite different.

Author: I'm just going to remove the last sentence but maybe let's address the existing JH integration on a case-by-case basis until we have a proper process?

every time you update your code.
You can use the `create_python_pipeline` function to create a pipeline from
within your JupyterHub Notebook and store your code locally rather than
in a Docker container.

If you want to create a pipeline from a Docker image or non-Python code, you
can use the
[create_pipeline](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.Client.create_pipeline)
method. This is the method Pachyderm uses when you run `pachctl create
pipeline` or create a pipeline in the Pachyderm UI. When you use
`create_pipeline`, you need to build a new Docker image and push
it to an image registry every
time you update the code in your pipeline. Users that are less familiar
with Docker might find this process a bit cumbersome.

When you use JupyterHub, in addition to the mentioned `create_pipeline` method,
you can use the [create_python_pipeline](https://pachyderm.github.io/python-pachyderm/python_pachyderm.m.html#python_pachyderm.create_python_pipeline) function that does not require
you to include your code in a Docker image and rebuild it each time you make
a change. Instead, this function creates a PFS repository

> called `<pipeline_name>.source` and puts the source code into it. Also, it

Reviewer 2: This is still incorrectly using a dot rather than an score (same with the following line):

```suggestion
called `<pipeline_name>_source` and puts the source code into it. Also, it
```

Author: added
